File: python/ray/_private/runtime_env_prototype/pluggability/plugin_schema_manager.py
import jsonschema
import logging
from typing import List
import json

logger = logging.getLogger(__name__)


class RuntimeEnvPluginSchemaManager:
    """ This mananger is used to load plugin json schemas.
    """
    schemas = {}

    @classmethod
    def load_schemas(cls, schema_paths: List[str]):
        """ """
        for schema_path in schema_paths:
            schema = json.load(open(schema_path))
            if "title" not in schema:
                logger.error("No valid title in %s", schema_path)
                continue
            if schema["title"] in cls.schemas:
                logger.error(
                    "The schema %s 'title' conflicts with %s",
                    schema_path,
                    cls.schemas[schema["title"]],
                )
                continue
            cls.schemas[schema["title"]] = schema

    @classmethod
    def validate(cls, title, instance):
        if title not in cls.schemas:
            raise TypeError(f"Invalid plugin name {title}")
        jsonschema.validate(instance=instance, schema=cls.schemas[title])

File: python/ray/_private/runtime_env_prototype/sdk/java/src/main/java/io/ray/PluginSchemaManager.java
package io.ray;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;


import com.fasterxml.jackson.databind.JsonNode;
import com.github.fge.jackson.JsonLoader;
import com.github.fge.jsonschema.core.report.ProcessingReport;
import com.github.fge.jsonschema.main.JsonSchemaFactory;
import com.github.fge.jsonschema.main.JsonValidator;

public class PluginSchemaManager {
  private final JsonValidator VALIDATOR = JsonSchemaFactory.byDefault().getValidator();
  private Map<String, JsonNode> schemas = new HashMap<>();
  private PluginSchemaManager(){ }

  private static class holder {
    private static final PluginSchemaManager INSTANCE = new PluginSchemaManager();
  }

  public static PluginSchemaManager getInstance() {
    return holder.INSTANCE;
  }

  public void loadSchemas(String[] schemaPaths) throws IOException {
    for (String path : schemaPaths) {
      JsonNode schema = JsonLoader.fromPath(path);
      JsonNode title = schema.get("title");
      if (title == null) {
        System.err.println("No valid title in " + path);
        continue;
      }
      String name = title.asText();
      if (schemas.containsKey(name)) {
        System.err.println("The schema " + path + "'title' already exists");
        continue;
      }
      schemas.put(name, schema);
    }
  }

    public void validate(String name, JsonNode jsonNode) throws Exception {
      if (!schemas.containsKey(name)) {
        throw new Exception("Invalid plugin name " + name);
      }
      ProcessingReport result = VALIDATOR.validate(schemas.get(name), jsonNode, true);
      if (!result.isSuccess()) {
        throw new Exception("Validate failed: " + result.toString());
      }
  }

}
